### PR TITLE
Query: Adds Correlated ActivityId wiring through query and bumps Direct version

### LINKED ISSUE #2391: Query correlated activity id
The query backend service can now handle correlated activity id. This will enable better troubleshooting of query issues as there currently isn't any way to correlate cross partition queries.

The change that need to be made:
1. New query should create a new correlated activity id. All the query request sent to the backend should include this property in the header with the key `"x-ms-cosmos-correlated-activityid"` is in the `WFConstants.CorrelatedActivityId`
2. The ITrace diagnostics need to include the correlated activity id
3. The continuation token should be modified to include the correlated activity id. This can be done in a separate PR. Adding the property should not break back compatibility of the current token.

Query already has a [partially wired through correlated activity id](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/12c8affe6119bbeb2410acb4a635b4b4b6622aee/Microsoft.Azure.Cosmos/src/Query/v3Query/QueryIterator.cs#L78). This will need to be expanded to send it through the NetworkAttachedDocumentContainer and the [query client](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/12c8affe6119bbeb2410acb4a635b4b4b6622aee/Microsoft.Azure.Cosmos/src/Query/Core/QueryClient/CosmosQueryClient.cs#L50).

Testing should use a custom requesthandler to make sure that all the pages sent by the query include the correct correlated activity id.

### PR DESCRIPTION
**Pull request by asketagarwal**
# Pull Request Template

## Description

This PR enable the SDK to send a common correlated activity id to BE, which will lead to better troubleshooting.

## Closing issues

closes #2391

### CODE REVIEW

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosBasicQueryTests.cs [hunk 1]
```diff
@@ -685,6 +687,74 @@ public async Task ConclictsTests(bool directMode)
...
+            {
+                if (requestMessage.OperationType == Documents.OperationType.Query)
+                {
+                    bool headerPresent = requestMessage.Headers.TryGetValue(Microsoft.Azure.Documents.WFConstants.BackendHeaders.CorrelatedActivityId, out string requestActivityId);
+                    if (!headerPresent)
+                    {
+                        Assert.Fail("Correlated ActivityId header not present in request");
+                    }
+
+                    if (this.correlatedActivityId == null)
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosBasicQueryTests.cs:743]: Add a check for Guid.Empty

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosBasicQueryTests.cs [hunk 2]
```diff
@@ -685,6 +687,74 @@ public async Task ConclictsTests(bool directMode)
...
+                                                                      "/pk",
+                                                                      throughput: 12000);
+
+            // Create items
+            for (int i = 0; i < 500; i++)
+            {
+                await container.CreateItemAsync<ToDoActivity>(ToDoActivity.CreateRandomToDoActivity());
+            }
+
+            FeedIterator<ToDoActivity> feedIterator = container.GetItemQueryIterator<ToDoActivity>(
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/CosmosBasicQueryTests.cs:711]: Set a page limit to verify that it always to verify it gets sent across multiple pages.

File: Microsoft.Azure.Cosmos/src/Query/v3Query/QueryIterator.cs [hunk 3]
```diff
@@ -75,11 +79,12 @@ internal sealed class QueryIterator : FeedIteratorInternal
                 resourceLink: resourceLink,
                 isContinuationExpected: isContinuationExpected,
                 allowNonValueAggregateQuery: allowNonValueAggregateQuery,
-                correlatedActivityId: Guid.NewGuid());
+                correlatedActivityId: correlatedActivityId);
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/v3Query/QueryIterator.cs:82]: This needs to be included in the ITrace diagnostics.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Query/SanityQueryTests.cs [hunk 4]
```diff
@@ -1006,8 +1006,11 @@ static async Task ImplementationAsync(Container container, IReadOnlyList<CosmosO
                     {
                         while (feedIterator.HasMoreResults)
                         {
-                            await feedIterator.ReadNextAsync(trace, default);
-                            numChildren++;
+                            using (ITrace childTrace = trace.StartChild("ChildTrace"))
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Query/SanityQueryTests.cs:1009]: Why is this needed?

**Comment by FabianMeiswinkel** [comment]
Please test this end-to-end at least manually. From my understanding the CorrelatedActivityId isn't wired up properly in the Rntbd transport yet - and it would never be actually sent to the backend at least when using direct transport . This feature is only useful when the header is actually sent to the backend (and makes it into Kusto SqlQueryExecMetric table)

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Query/SanityQueryTests.cs (hunk 4 above)
Comment (asketagarwal) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Query/SanityQueryTests.cs:1038]: This test is trying to add readnextasync's traces to a single root trace, which duplicates the correlatedactivityid and leads to an error.

Comment (j82w) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Query/SanityQueryTests.cs:1038]: Is there any way to avoid this exception? That seems like a valid scenario where we want a single root trace to have multiple readnextasync calls.

**Review by kirankumarkolli — APPROVED** [review]

**Review by j82w — APPROVED** [review]

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Tracing/EndToEndTraceWriterBaselineTests.cs [hunk 5]
```diff
@@ -1,1432 +1,1443 @@
...
-            {
-                await Task.Delay(this.timeToSleep);
-                return await base.SendAsync(request, cancellationToken);
-            }
-        }
-    }
-}
+﻿namespace Microsoft.Azure.Cosmos.EmulatorTests.Tracing
+{
+    using System;
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/Tracing/EndToEndTraceWriterBaselineTests.cs]: Did the line endings get changed? Please revert all the white space changes it breaks the git blame.

**Review by j82w — APPROVED** [review]

File: Directory.Build.props [hunk 6]
```diff
@@ -4,7 +4,7 @@
 		<ClientOfficialVersion>3.23.0</ClientOfficialVersion>
 		<ClientPreviewVersion>3.23.0</ClientPreviewVersion>
 		<ClientPreviewSuffixVersion>preview</ClientPreviewSuffixVersion>
-		<DirectVersion>3.23.1</DirectVersion>
+		<DirectVersion>3.24.1</DirectVersion>
```
Comment (j82w) [inline Directory.Build.props:7]: The 3.24.1 bump should be done in a separate PR to keep a clear git history. 3.24.1 adds a lot of changes and improvements to the HA and diagnostics. Merging those changes into the query correlated activity id will hide those other improvements.

Comment (asketagarwal) [reply Directory.Build.props]: Reset this PR. Will create a new PR for updating Direct version

**Review by j82w — APPROVED** [review]
